cmd/data-crawler.go

```diff
@@ -876,26 +873,41 @@ func (i *crawlItem) applyActions(ctx context.Context, o ObjectLayer, meta action
 			if !obj.DeleteMarker { // if this is not a delete marker log and return
 				// Do nothing - heal in the future.
 				logger.LogIf(ctx, err)
-				return size
+				return false, size
 			}
 		case ObjectNotFound, VersionNotFound:
 			// object not found or version not found return 0
-			return 0
+			return false, 0
 		default:
 			// All other errors proceed.
 			logger.LogIf(ctx, err)
-			return size
+			return false, size
 		}
 	}
 
-	var applied bool
 	action = evalActionFromLifecycle(ctx, *i.lifeCycle, obj, i.debug)
 	if action != lifecycle.NoneAction {
 		applied = applyLifecycleAction(ctx, action, o, obj)
 	}
 
 	if applied {
-		return 0
+		return true, 0
+	}
+
+	return false, size
+}
+
+// applyActions will apply lifecycle checks on to a scanned item.
+// The resulting size on disk will always be returned.
+// The metadata will be compared to consensus on the object layer before any changes are applied.
+// If no metadata is supplied, -1 is returned if no action is taken.
+func (i *crawlItem) applyActions(ctx context.Context, o ObjectLayer, meta actionMeta) int64 {
+	applied, size := i.applyLifecycle(ctx, o, meta)
+	// For instance, an applied lifecycle means we remove/transitioned an object
+	// from the current deployment, which means we don't have to call healing
+	// routine even if we are asked to do via heal flag.
+	if !applied && i.heal {
```

Comment:
if it applies lifecycle to only create a deletemarker to expire the current version of object  will `applied` be set to `true`? If yes, it still needs to be healed.

Reply:
> if it applies lifecycle to only create a deletemarker to expire the current version of object will `applied` be set to `true`? If yes, it still needs to be healed.

@kannappanr lifecycle does never touch any latest version, like a delete marker in your use case, so applied will always be false in this case. The delete marker will be healed later by self healing or mc admin heal.